internal/pkg/agent/application/gateway/fleet/fleet_gateway.go

```diff
@@ -155,38 +179,156 @@ func (f *fleetGateway) Run(ctx context.Context) error {
 			return ctx.Err()
 		case <-f.scheduler.WaitTick():
 			f.log.Debug("FleetGateway calling Checkin API")
+			func() {
+				// get current state
+				stateUpdateSubCtx, cancelStateUpdateSub := context.WithCancel(ctx)
+				stateUpdateCh := f.stateFetcher.StateSubscribe(stateUpdateSubCtx).Ch()
+				defer func() {
+					cancelStateUpdateSub()
+				}()
+				state := <-stateUpdateCh
+
+				var checkinResponse *fleetapi.CheckinResponse
+
+				for checkinResponse == nil {
+					if ctx.Err() != nil {
+						return
+					}
+
+					checkinResult := f.performCancellableCheckin(ctx, state, stateUpdateCh, backoff, f.debouncerFactory())
+
+					if checkinResult.err != nil {
+						var relaunchCheckinErr *needNewCheckinError
+						if errors.As(checkinResult.err, &relaunchCheckinErr) {
+							// checkin was cancelled because we have an updated state to send to fleet
+							state = relaunchCheckinErr.newState
+						}
+
+						continue
+					}
+
+					// if we got here the checkin has been completed
+					checkinResponse = checkinResult.response
+				}
+
+				f.log.Debugf("Processing checkin response %v", checkinResponse)
```

Comment:
That is a debug log that I needed for debugging the agent while developing the async checkin (mostly to understand that we got some non-error answer and we are going to process the actions)

Since we can have more or less the same information by specifiying `DEBUG_AGENT=1` in env I am going to remove the checking response value from here ;)